### DISCUSSION TITLE
怎么使用generate打包后的文件源码中包含后端请求回来的数据？

### DISCUSSION THREAD

**Discussion by w569638598** [nuxt/nuxt#30294] [Questions]
使用generate打包成html文件后，但是我想在比如我的文章详情中，我需要显示从后端请求回来的数据在网页源码中。因为要seo，
![image](https://github.com/user-attachments/assets/12ed30c0-e34a-46b6-a797-9a4864c9f19f)
这样请求的数据，在setup中。我打generate包只有，有的
![image](https://github.com/user-attachments/assets/b8aceda1-d5e7-4712-b29e-d4ed70bc371e)
文章详情打包后有这个单独的页面，有的没有，所以我不太清楚到底逻辑是什么？请求帮助啊。英语不是很好。可以说很烂。

**Comment by xjccc** [comment]
If possible some `ids` have no result?

**Reply by w569638598** [reply]
不，每个都有数据，这些是我们后台添加的文章。数据都没问题。只是打包后不知道为什么没有对应id的数据。并且我不想每个文章都打包成单独的页面。我想在服务器端请求后，直接js生成html文件内容。

**Reply by xjccc** [reply]
I think you should use `ssr`. Not `generate html`

**Reply by w569638598** [reply]
但是不行啊。后端不想动服务器，服务器是contos7，安装node18+的话，他们说需要contos8，所以他们嫌麻烦。

**Reply by xjccc** [reply]
Use `pnpm generate`, will generate all html, and you should upload all your html to server. Then you could access the html from server. If you don't have corresponding html, user can't access it.

**Reply by w569638598** [reply]
所以我的这种需求没办法实现吗？‘我不想每个文章都打包成单独的页面。我想在服务器端请求后，直接js生成html文件内容。’或者‘打包后都有对应的文章id的html页面。’现在是有的有，有的没有。

**Reply by xjccc** [reply]
If you want generate html in server, `pnpm generate` is not suitable. 

You could create a mini repo for more details. Picture is not clearly.

**Reply by w569638598** [reply]
https://stackblitz.com/edit/nuxt-starter-c7opuozq?file=dist
你看这个例子，打包之后，并没有detail对应的id页面，点击查看网页源码，也没有对应的数据，哪怕我详情数据不是从后台取得。我不是很懂。

**Reply by xjccc** [reply]
https://nuxt.com/docs/api/nuxt-config#generate
You could see this.

**Reply by w569638598** [reply]
我看过这个，但是感觉不太行，我有1000万篇文章，难道我这样生成1000万个路由吗？肯定不能把所有的文章都请求回来，用id生成路由。

**Reply by w569638598** [reply]
还有个问题，我使用generate打包后，详情页面没有根据文章id自动生成对应的html页面对吧？然后我打包后，从列表进没问题，但是在浏览器刷新（F5）就会报404。
![image](https://github.com/user-attachments/assets/91f00bd7-15e9-4ea8-b46f-2f9b6b5ca7bb)
这种怎么解决。

**Reply by xjccc** [reply]
你这个是因为服务端没有history路由的转发，需要设置一下

https://router.vuejs.org/guide/essentials/history-mode.html#nginx

**Reply by xjccc** [reply]
生成静态页，就跟php模版一样，你需要把静态生成的html放在服务器上，才能访问。如果你想服务端生成，那你得用`ssr`这个跨不过去。你如果只部署`index.html`，那就是`spa`模式

**Reply by w569638598** [reply]
就我这样的需求的，有么有什么例子我学习下啊大哥。

**Reply by xjccc** [reply]
没。。现在nuxt最低支持18+吧。后面再升级可能需要node版本更高。最好还是服务端渲染 + CDN服务。很方便了，在者就是`vue spa`开发也很方便

**Comment by itshizhan** [comment]
你这不是spa的效果吗，直接vue啊，干啥使用nuxt

**Reply by w569638598** [reply]
不是spa啊。好像也是，反正我的需求就是，访问的页面中都是渲染好的代码。比如详情是已经渲染好的文章内容的html页面，就是查看网页源码时，代码里面是包含请求回来的数据的，且每个页面都有自己的title，keywords什么的，因为需要seo。
但是你说的vue和spa，我理解的是一个静态模板，body里面内容是根据假路由渲染的对应的component。

**Reply by youwei997** [reply]
> 不是spa啊。好像也是，反正我的需求就是，访问的页面中都是渲染好的代码。比如详情是已经渲染好的文章内容的html页面，就是查看网页源码时，代码里面是包含请求回来的数据的，且每个页面都有自己的title，keywords什么的，因为需要seo。 但是你说的vue和spa，我理解的是一个静态模板，body里面内容是根据假路由渲染的对应的component。

你想要的是不是博客的那种，点进去都是已经渲染好的静态页。你搜一下nuxt ssg，看看是不是你想要的效果

**Reply by w569638598** [reply]
是ssg，但是一直找不到到底怎么用的，我就会npm run generate。搜也搜不到怎么用的。你那有例子没？

**Comment by nowo** [comment]
你的需求就是要做SEO，那就直接在服务器上运行项目呗！用`pm2`做个进程守护

**Reply by nowo** [reply]
最后nginx指向这个运行的地址，外部就能正常访问了！

**Reply by w569638598** [reply]
后端一下不懂的情况下，只是一个静态服务。没有什么代理。PM2什么的。前端把某些页面直接打成html文件，文章详情这样的不打html。然后我想文章详情这样的要用完全后端请求渲染返回html。

**Reply by nowo** [reply]
> 后端一下不懂的情况下，只是一个静态服务。没有什么代理。PM2什么的。前端把某些页面直接打成html文件，文章详情这样的不打html。然后我想文章详情这样的要用完全后端请求渲染返回html。

那你这样就是SSG了，但是你项目上线后，如果你后台又添加了一条新文章，这个是访问不到的。
